scripts/check_compliance.py

```diff
@@ -426,73 +426,123 @@ class Codeowners(ComplianceTest):
+        """Appends and prepends '**[/*]' when needed. Result has neither a
+        leading nor a trailing slash.
+        """
+
+        ret = git_pattern
+
+        if git_pattern.startswith("/"):
+            ret = ret[1:]
+        else:
+            ret = "**/" + ret
+
+        if git_pattern.endswith("/"):
+            ret = ret + "**/*"
+        elif os.path.isdir(os.path.join(git_top, ret)):
+            self.add_failure("Expected '/' after directory '{}' "
+                             "in CODEOWNERS".format(ret))
+
+        return ret
+
 
     def run(self):
         self.prepare()
-        git_root = sh.git("rev-parse", "--show-toplevel").strip()
-        codeowners = os.path.join(git_root, "CODEOWNERS")
+        # TODO: testing an old self.commit range that doesn't end
+        # with HEAD is most likely a mistake. Should warn, see
+        # https://github.com/zephyrproject-rtos/ci-tools/pull/24
+        codeowners = os.path.join(git_top, "CODEOWNERS")
         if not os.path.exists(codeowners):
             self.skip("CODEOWNERS not available in this repo")
 
         commit = sh.git("diff","--name-only", "--diff-filter=A", self.commit_range, **sh_special_args)
-        new_files = commit.split("\n")
-        files_in_tree = sh.git("ls-files",  **sh_special_args).split("\n")
-        if new_files:
-            owned = self.parse_codeowners(git_root, codeowners)
-            new_not_owned = []
-            for f in new_files:
-                if not f:
-                    continue
-                if f not in owned:
-                    new_not_owned.append(f)
-
-            if new_not_owned:
-                self.add_failure("New files added that are not covered in "
-                                 "CODEOWNERS:\n\n" + "\n".join(new_not_owned) +
-                                 "\n\nPlease add one or more entries in the "
-                                 "CODEOWNERS file to cover those files")
+        new_files = commit.splitlines()
+        logging.debug("New files %s" % new_files)
+
+        if not new_files:
+            # TODO: parse CODEOWNERS to report errors in it *without*
+            # scanning the tree to keep this case very fast.
+            return
+
+        # Convert to pathlib.Path string representation (e.g.,
+        # double-backslashes 'dir1\\dir2\\' on Windows) to be consistent
+        # with self.ls_owned_files()
+        new_files = [ str(Path(f)) for f in new_files ]
+
+        logging.info("If this takes too long then cleanup and try again")
+        patrn2files = self.ls_owned_files(codeowners)
+
+        new_not_owned = []
+        for f in new_files:
+            f_is_owned = False
+
+            for git_pat, owned in patrn2files.items():
+                logging.debug("Scanning %s for %s" % (git_pat, f))
+
+                if f in owned:
+                    logging.info("%s matches new file %s", git_pat, f)
+                    f_is_owned = True
+                    break
```

Comment:
Maybe it's harder to read for people who don't do that much Python, but another option is to use a `for ... else`:

```python
for git_pat, owned in patrn2files.items():
    ...
    if new_file in owned:
        ...
        break
else:
    new_not_owned.append(new_file)
```

Reply:
Agree this is a perfect use case for this "for/else" syntax. Also agree this syntax is very unusual, hence harder to read and would require almost as much comment as the amount of code it would save. Guido wrote he regrets this feature: https://intoli.com/blog/for-else-in-python/
Any chance it would help with performance? If not then I'd rather not.

Reply:
Fine with me to skip it.